File: sections/20_zero_one_bfs_variants/problems/c_warp_maze/solution.py
from collections import deque
import sys


def main() -> None:
    input = sys.stdin.readline
    n, m = map(int, input().split())
    grid = [list(input().strip()) for _ in range(n)]
    start = goal = (-1, -1)
    for r in range(n):
        for c in range(m):
            if grid[r][c] == "S":
                start = (r, c)
            elif grid[r][c] == "G":
                goal = (r, c)
    inf = 10**9
    dist = [[inf] * m for _ in range(n)]
    sr, sc = start
    dist[sr][sc] = 0
    dq = deque([start])
    while dq:
        r, c = dq.popleft()
        d = dist[r][c]
        for dr, dc in ((1,0),(-1,0),(0,1),(0,-1)):
            nr, nc = r + dr, c + dc
            if 0 <= nr < n and 0 <= nc < m and grid[nr][nc] != "#" and d < dist[nr][nc]:
                dist[nr][nc] = d
                dq.appendleft((nr, nc))
        for dr in range(-2, 3):
            for dc in range(-2, 3):
                nr, nc = r + dr, c + dc
                if 0 <= nr < n and 0 <= nc < m and grid[nr][nc] != "#" and d + 1 < dist[nr][nc]:
                    dist[nr][nc] = d + 1
                    dq.append((nr, nc))
    gr, gc = goal
    print(-1 if dist[gr][gc] == inf else dist[gr][gc])


if __name__ == "__main__":
    main()

File: sections/20_zero_one_bfs_variants/problems/c_warp_maze/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, m;
    if (!(cin >> n >> m)) return 0;
    vector<string> grid(n);
    pair<int,int> start{-1,-1}, goal{-1,-1};
    for (int r = 0; r < n; r++) {
        cin >> grid[r];
        for (int c = 0; c < m; c++) {
            if (grid[r][c] == 'S') start = {r,c};
            if (grid[r][c] == 'G') goal = {r,c};
        }
    }
    const int INF = 1e9;
    vector<vector<int>> dist(n, vector<int>(m, INF));
    deque<pair<int,int>> dq;
    dist[start.first][start.second] = 0;
    dq.push_back(start);
    int wr[4] = {1,-1,0,0}, wc[4] = {0,0,1,-1};
    while (!dq.empty()) {
        auto [r,c] = dq.front(); dq.pop_front();
        int d = dist[r][c];
        for (int i = 0; i < 4; i++) {
            int nr = r + wr[i], nc = c + wc[i];
            if (nr < 0 || nr >= n || nc < 0 || nc >= m || grid[nr][nc] == '#') continue;
            if (d < dist[nr][nc]) {
                dist[nr][nc] = d;
                dq.push_front({nr,nc});
            }
        }
        for (int dr = -2; dr <= 2; dr++) {
            for (int dc = -2; dc <= 2; dc++) {
                int nr = r + dr, nc = c + dc;
                if (nr < 0 || nr >= n || nc < 0 || nc >= m || grid[nr][nc] == '#') continue;
                if (d + 1 < dist[nr][nc]) {
                    dist[nr][nc] = d + 1;
                    dq.push_back({nr,nc});
                }
            }
        }
    }
    int ans = dist[goal.first][goal.second];
    cout << (ans == INF ? -1 : ans) << '\n';
    return 0;
}

